File: app/src/main/java/cat/xojan/random1/data/PreferencesDownloadPodcastRepository.java
package cat.xojan.random1.data;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.Nullable;

import java.util.HashSet;
import java.util.Set;

import cat.xojan.random1.domain.entities.Podcast;
import cat.xojan.random1.domain.repository.DownloadPodcastRepository;

public class PreferencesDownloadPodcastRepository implements DownloadPodcastRepository {

    private static final String DOWNLOAD_PODCASTS = "dowload_podcasts_repo";
    private static final String DOWNLOADING_PODCASTS = "downloading_podcasts";
    private static final String DOWNLOADED_PODCASTS = "downloaded_podcasts";

    private final Gson mGson;
    private SharedPreferences mPreferences;

    public PreferencesDownloadPodcastRepository(Context context) {
        mPreferences = context.getSharedPreferences(DOWNLOAD_PODCASTS, Context.MODE_PRIVATE);
        mGson = new Gson();
    }

    @Override
    public boolean addDownloadingPodcast(Podcast podcast) {
        Set<Podcast> podcasts = getDownloadingPodcasts();
        podcast.setState(Podcast.State.DOWNLOADING);
        podcasts.add(podcast);
        return mPreferences.edit()
                .putString(DOWNLOADING_PODCASTS, setToJson(podcasts))
                .commit();
    }

    @Override
    public boolean deleteDownloadingPodcast(Podcast podcast) {
        Set<Podcast> podcasts = getDownloadingPodcasts();
        podcasts.remove(podcast);
        return mPreferences.edit()
                .putString(DOWNLOADING_PODCASTS, setToJson(podcasts))
                .commit();
    }

    @Override
    public void setPodcastAsDownloaded(String audioId, String filePath) {
        Podcast podcast = getDownloadingPodcast(audioId);
        if (podcast != null && deleteDownloadingPodcast(podcast)) {
            podcast.setFilePath(filePath);
            podcast.setState(Podcast.State.DOWNLOADED);
            addDownloadedPodcast(podcast);
        }
    }

    public Set<Podcast> getDownloadingPodcasts() {
        String jsonPodcasts = mPreferences.getString(DOWNLOADING_PODCASTS, "");
        return jsonToSet(jsonPodcasts);
    }

    public Set<Podcast> getDownloadedPodcasts() {
        String jsonPodcasts = mPreferences.getString(DOWNLOADED_PODCASTS, "");
        return jsonToSet(jsonPodcasts);
    }

    @Override
    public void deleteDownloadedPodcast(Podcast podcast) {
        Set<Podcast> podcasts = getDownloadedPodcasts();
        podcasts.remove(podcast);
        mPreferences.edit()
                .putString(DOWNLOADED_PODCASTS, setToJson(podcasts))
                .apply();
    }

    @Override @Nullable
    public String getDownloadedPodcastTitle(String audioId) {
        Set<Podcast> podcasts = getDownloadedPodcasts();
        for (Podcast podcast : podcasts) {
            if (podcast.getAudioId().equals(audioId)) {
                return podcast.getTitle();
            }
        }
        return null;
    }

    private void addDownloadedPodcast(Podcast podcast) {
        Set<Podcast> podcasts = getDownloadedPodcasts();
        podcasts.add(podcast);
        mPreferences.edit()
                .putString(DOWNLOADED_PODCASTS, setToJson(podcasts))
                .apply();
    }

    @Nullable
    private Podcast getDownloadingPodcast(String audioId) {
        Podcast downloadedPodcast = null;

        for (Podcast podcast : getDownloadingPodcasts()) {
            if (podcast.getAudioId().equals(audioId)) {
                downloadedPodcast = podcast;
                break;
            }
        }
        return downloadedPodcast;
    }

    private Set<Podcast> jsonToSet(String json) {
        TypeToken<Set<Podcast>> token = new TypeToken<Set<Podcast>>() {};
        Set<Podcast> podcasts = mGson.fromJson(json, token.getType());
        return podcasts == null ? new HashSet<Podcast>() : podcasts;
    }

    private String setToJson(Set<Podcast> podcasts) {
        return mGson.toJson(podcasts);
    }
}

File: app/src/main/java/cat/xojan/random1/data/PreferencesDownloadPodcastRepository.kt
package cat.xojan.random1.data

import android.content.Context
import android.content.SharedPreferences
import cat.xojan.random1.domain.entities.Podcast
import cat.xojan.random1.domain.repository.DownloadPodcastRepository
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import com.squareup.moshi.Rfc3339DateJsonAdapter
import com.squareup.moshi.Types
import java.util.*

class PreferencesDownloadPodcastRepository(context: Context) : DownloadPodcastRepository {

    private val DOWNLOAD_PODCASTS = "dowload_podcasts_repo"
    private val DOWNLOADING_PODCASTS = "downloading_podcasts"
    private val DOWNLOADED_PODCASTS = "downloaded_podcasts"

    private val sharedPref: SharedPreferences
    private val jsonAdapter: JsonAdapter<MutableSet<Podcast>>

    init {
        sharedPref = context.getSharedPreferences(DOWNLOAD_PODCASTS, Context.MODE_PRIVATE)
        val type = Types.newParameterizedType(MutableSet::class.java, Podcast::class.java)
        val moshi = Moshi.Builder()
                .add(Date::class.java, Rfc3339DateJsonAdapter().nullSafe())
                .build()
        jsonAdapter = moshi.adapter(type)
    }

    override fun addDownloadingPodcast(podcast: Podcast): Boolean {
        podcast.state = Podcast.State.DOWNLOADING
        val podcasts = getDownloadingPodcasts()
        podcasts.add(podcast)
        return sharedPref.edit()
                .putString(DOWNLOADING_PODCASTS, setToJson(podcasts))
                .commit()
    }

    override fun deleteDownloadingPodcast(podcast: Podcast): Boolean {
        val podcasts = getDownloadingPodcasts()
        podcasts.remove(podcast)
        return sharedPref.edit()
                .putString(DOWNLOADING_PODCASTS, setToJson(podcasts))
                .commit()
    }

    override fun setPodcastAsDownloaded(audioId: String, filePath: String) {
        val podcast = getDownloadingPodcast(audioId)
        if (podcast != null && deleteDownloadingPodcast(podcast)) {
            podcast.filePath = filePath
            podcast.state = Podcast.State.DOWNLOADED
            addDownloadedPodcast(podcast)
        }
    }

    override fun getDownloadingPodcasts(): MutableSet<Podcast> {
        val podcastsJson = sharedPref.getString(DOWNLOADING_PODCASTS, null)
        return jsonToSet(podcastsJson)
    }

    override fun getDownloadedPodcasts(): MutableSet<Podcast> {
        val podcastsJson = sharedPref.getString(DOWNLOADED_PODCASTS, null)
        return jsonToSet(podcastsJson)
    }

    override fun deleteDownloadedPodcast(podcast: Podcast) {
        val podcasts = getDownloadedPodcasts()
        podcasts.remove(podcast)
        sharedPref.edit()
                .putString(DOWNLOADED_PODCASTS, setToJson(podcasts))
                .apply()
    }

    override fun getDownloadedPodcastTitle(audioId: String): String? {
        return getDownloadedPodcasts()
                .firstOrNull { it.audioId == audioId }
                ?.title
    }

    private fun jsonToSet(json: String?): MutableSet<Podcast> {
        if (json == null) {
            return mutableSetOf()
        }
        val podcasts: MutableSet<Podcast>? = jsonAdapter.fromJson(json)
        return podcasts ?: mutableSetOf()
    }

    private fun setToJson(podcasts: MutableSet<Podcast>): String {
        return jsonAdapter.toJson(podcasts)
    }

    private fun getDownloadingPodcast(audioId: String): Podcast? {
        return getDownloadingPodcasts().firstOrNull { it.audioId == audioId }
    }

    private fun addDownloadedPodcast(podcast: Podcast) {
        val podcasts = getDownloadedPodcasts()
        podcasts.add(podcast)
        sharedPref.edit()
                .putString(DOWNLOADED_PODCASTS, setToJson(podcasts))
                .apply()
    }
}